### PR TITLE
ExCoT-DPO project

### PR DESCRIPTION
**Pull request by sfc-gh-bzhai**
TODO:
- [x] Validate implementation via training loss curve
- [x] Remove extra dpo example used in testing
- [x] Ensure all config fields are used and have doc strings
- [x] Test synth data generation saving and loading for training
- [x] Add quickstart readme on how to generate data, training using DPO trainer, and evaluation

### CODE REVIEW

File: arctic_training/data/dpo_factory.py [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+
+    mask_inputs: bool = True
+    """ Whether to mask the input sequence. """
+
+    always_max_length: bool = False
+    """
+    If this is turned on, each batch will be filled up to the max length by
+    appending samples until the total length matches the max length. It might
+    cause the last sample to be truncated.
+    """
```
Comment (sfc-gh-mwyatt) [inline arctic_training/data/dpo_factory.py:54]:
```suggestion
    """
    
```

File: arctic_training/data/dpo_factory.py [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+    dpo_prompt_truncation_mode: str = "keep_start"
```
Comment (sfc-gh-mwyatt) [inline arctic_training/data/dpo_factory.py:55]: Can we add a comment for autogen docs to this field?

File: arctic_training/data/dpo_factory.py [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+            rejected_prompt_len_input_ids == 0
+            or bos_token_id != rejected_tokens["prompt_input_ids"][0]
+        ):
+            rejected_tokens["prompt_input_ids"] = [bos_token_id] + rejected_tokens[
+                "prompt_input_ids"
+            ]
+            rejected_tokens["prompt_attention_mask"] = [1] + rejected_tokens[
+                "prompt_attention_mask"
+            ]
+    return prompt_tokens, chosen_tokens, rejected_tokens
```
Comment (sfc-gh-mwyatt) [inline arctic_training/data/dpo_factory.py:128]:
```suggestion
def add_bos_token_if_needed(
    bos_token_id: Union[None, int],
    tokens: Dict[str, List[int]],
) -> Dict[str, List[int]]:
    len_input_ids = len(tokens["prompt_input_ids"])
    if bos_token_id is not None and (len_input_ids == 0 or bos_token_id != tokens["prompt_token_ids"][0]):
                tokens["prompt_input_ids"].insert(0, bos_token_id)
                tokens["prompt_attention_mask].insert(0, 1)
    return tokens
```

This could be more concise. And then to call the function:
```python
prompt_tokens = add_bos_token_if_needed(tokenizer.bos_token_id, prompt_tokens)
chosen_tokens = add_bos_token_if_needed(tokenizer.bos_token_id, chosen_tokens)
rejected_tokens = add_bos_token_if_needed(tokenizer.bos_token_id, rejected_tokens)
```

File: arctic_training/data/dpo_factory.py [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+                "attention_mask": attention_mask,
+                "prompt": [example["prompt_input_ids"] for example in instances]
+                "prompt_text": prompt_text,
+                "chosen_text": chosen_text,
+                "rejected_text": rejected_text,
+            }
+        return rt
+
+
+@register
```
Comment (sfc-gh-mwyatt) [inline arctic_training/data/dpo_factory.py:183]:
```suggestion
```

No need to explicitly call register decorator since merging of #56

File: arctic_training/data/dpo_factory.py [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+                "Dataset must have 'prompt' column to tokenize for SFTDataFactory."
+            )
+        if "chosen" not in dataset.column_names:
+            raise ValueError(
+                "Dataset must have 'chosen' column to tokenize for SFTDataFactory."
+            )
+        if "rejected" not in dataset.column_names:
+            raise ValueError(
+                "Dataset must have 'rejected' column to tokenize for SFTDataFactory."
+            )
```
Comment (sfc-gh-mwyatt) [inline arctic_training/data/dpo_factory.py:206]:
```suggestion
        missing_columns = [c for c in ("prompt", "chosen", "rejected") if c not in dataset.column_names]
        if len(missing_columns) > 0:
            raise ValueError(
                f"Dataset must have 'prompt', 'chosen', and 'rejected' columns to tokenizer for DPODataFactory. Missing the following columns: {missing_columns}"
            )
```

This way we can have an error message that indicate if more than one column is missing from the incoming dataset.

File: arctic_training/trainer/dpo_trainer.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+class DPOTrainer(Trainer):
+    name = "dpo"
+    config: DPOTrainerConfig
+    model_factory: Union[HFModelFactory, LigerModelFactory]
+    ref_model_factory: Union[HFModelFactory, LigerModelFactory]
+    checkpoint_engine: Union[DSCheckpointEngine, HFCheckpointEngine]
+    optimizer_factory: Union[FusedAdamOptimizerFactory]
+    scheduler_factory: Union[HFSchedulerFactory]
+    tokenizer_factory: Union[HFTokenizerFactory]
```
Comment (sfc-gh-mwyatt) [inline arctic_training/trainer/dpo_trainer.py:46]:
```suggestion
    optimizer_factory: FusedAdamOptimizerFactory
    scheduler_factory: HFSchedulerFactory
    tokenizer_factory: HFTokenizerFactory
```

No need for union, we support both single type hints and unions now!

File: arctic_training/trainer/dpo_trainer.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        # dummy token; we'll ignore the losses on these tokens later
+        labels[labels == self.ignore_label_index] = 0
+
+        per_token_logps = torch.gather(
+            logits.log_softmax(-1), dim=2, index=labels.unsqueeze(2)
+        ).squeeze(2)
+
+        return (per_token_logps * loss_mask).sum(-1), loss_mask.sum(-1)
+
+    def init_ref_model(self):
```
Comment (sfc-gh-mwyatt) [inline arctic_training/trainer/dpo_trainer.py:82]: This method should be a function outside of the `DPOTrainer`

File: arctic_training/trainer/dpo_trainer.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    labels[labels == ignore_label_index] = 0
+
+    per_token_logps = torch.gather(
+        logits.log_softmax(-1), dim=2, index=labels.unsqueeze(2)
+    ).squeeze(2)
+
+    return (per_token_logps * loss_mask).sum(-1), loss_mask.sum(-1)
+
+class DPOTrainerConfig(TrainerConfig):
+    type: str = "dpo"
```
Comment (sfc-gh-mwyatt) [inline arctic_training/trainer/dpo_trainer.py:78]:
```suggestion
```

No need to set this. It will get automatically filled by pydantic validation

File: arctic_training/trainer/dpo_trainer.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    ref_model: ModelConfig
+    beta: float
+    ignore_label_index: int = -100
+    label_smoothing: float = 0.0
```
Comment (sfc-gh-mwyatt) [inline arctic_training/trainer/dpo_trainer.py:82]: Can you add some documentation strings?

File: arctic_training/trainer/dpo_trainer.py [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    checkpoint_engine: Union[DSCheckpointEngine, HFCheckpointEngine]
+    optimizer_factory: FusedAdamOptimizerFactory
+    scheduler_factory: HFSchedulerFactory
+    tokenizer_factory: HFTokenizerFactory
+    callbacks = [("post-init", init_ref_model)]
+
+
+    def init_ref_model(self):
+        ref_model_factory = self.config.ref_model.factory(trainer=self, config=self.config.ref_model) # Be explicit about which model config to use
+        self.ref_model = ref_model_factory()
```
Comment (sfc-gh-mwyatt) [inline arctic_training/trainer/dpo_trainer.py:100]:
```suggestion
def init_ref_model(self: Trainer):
    ref_model_factory = self.config.ref_model.factory(trainer=self, config=self.config.ref_model) # Be explicit about which model config to use
    self.ref_model = ref_model_factory()

class DPOTrainer(Trainer):
    name = "dpo"
    config: DPOTrainerConfig
    model_factory: Union[HFModelFactory, LigerModelFactory]
    ref_model_factory: Union[HFModelFactory, LigerModelFactory]
    checkpoint_engine: Union[DSCheckpointEngine, HFCheckpointEngine]
    optimizer_factory: FusedAdamOptimizerFactory
    scheduler_factory: HFSchedulerFactory
    tokenizer_factory: HFTokenizerFactory
    callbacks = [("post-init", init_ref_model)]
```

`init_ref_model` needs to be defined outside `DPOTrainer` or we will not know what this callback is referencing. The other option is to make it a static method of `DPOTrainer` and set the callback to `("post-init", DPOTrainer.init_ref_model)`

File: arctic_training/data/dpo_factory.py [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+
+
+class DPODataConfig(DataConfig):
+    max_length: int = 8192
+    """ Maximum length of the input sequence. """
+
+    max_prompt_length: int = 4096
+    """ Maximum prompt length of the input sequence. """
+
+    mask_inputs: bool = True
```
Comment (sfc-gh-mwyatt) [inline arctic_training/data/dpo_factory.py:42]: This field is never actually used anywhere. It's passed to `tokenize_messages`, but then never used. Can we either remove this or make sure that it is being used?

File: arctic_training/data/dpo_factory.py [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+    """ Whether to mask the input sequence. """
+
+    always_max_length: bool = False
```
Comment (sfc-gh-mwyatt) [inline arctic_training/data/dpo_factory.py:45]: This is also never used in the data factory.

File: arctic_training/data/dpo_factory.py [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+    cause the last sample to be truncated.
+    """
+
+    dpo_prompt_truncation_mode: str = "keep_start"
+
+
+def _adjust_prompt_length(
+    prompt_token,
+    chosen_token,
+    rejected_token,
```
Comment (sfc-gh-mwyatt) [inline arctic_training/data/dpo_factory.py:58]:
```suggestion
    prompt_token: List[int],
    chosen_token: List[int],
    rejected_token: List[int],
```
Add type hints

File: arctic_training/data/dpo_factory.py [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+    len_input_ids = len(tokens["prompt_input_ids"])
+    if bos_token_id is not None and (
+        len_input_ids == 0 or bos_token_id != tokens["prompt_input_ids"][0]
+    ):
+        tokens["prompt_input_ids"].insert(0, bos_token_id)
+        tokens["prompt_attention_mask"].insert(0, 1)
+    return tokens
+
+
+def _build_sequence_tokens(tokens, prefix: str) -> Dict:
```
Comment (sfc-gh-mwyatt) [inline arctic_training/data/dpo_factory.py:96]:
```suggestion
def _build_sequence_tokens(tokens: Dict[str, List[int]], prefix: str) -> Dict[str, List[int]]:
```
Better define type hints

File: arctic_training/data/dpo_factory.py [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+            "rejected_text": rejected_text,
+        }
+        return rt
+
+
+class DPODataFactory(DataFactory):
+    name = "dpo"
+    config: DPODataConfig
+
+    def convert_text(self, tokenizer, conversations: List[Dict[str, str]]):
```
Comment (sfc-gh-mwyatt) [inline arctic_training/data/dpo_factory.py:153]: Can we add a return type for this method?

File: arctic_training/trainer/dpo_trainer.py [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+    per_token_logps = torch.gather(
+        logits.log_softmax(-1), dim=2, index=labels.unsqueeze(2)
+    ).squeeze(2)
+
+    return (per_token_logps * loss_mask).sum(-1), loss_mask.sum(-1)
+
+
+class DPOTrainerConfig(TrainerConfig):
+    ref_model: ModelConfig
+    beta: float
```
Comment (sfc-gh-mwyatt) [inline arctic_training/trainer/dpo_trainer.py:85]: Should this have a default value?

File: arctic_training/trainer/dpo_trainer.py [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+        subconfig = cls._get_subconfig_object(
+            v=v,
+            info=info,
+            get_class_fn=get_registered_model_factory,
+            attr_name="ref_model_factory",
+        )
+        return cast(ModelConfig, subconfig)
+
+
+def init_ref_model(self: "DPOTrainer"):
```
Comment (sfc-gh-mwyatt) [inline arctic_training/trainer/dpo_trainer.py:104]:
```suggestion
def init_ref_model(self: "DPOTrainer") -> None:
```

File: examples/dpo.yaml
Comment (sfc-gh-mwyatt) [inline examples/dpo.yaml:1]: OK to keep this while we are doing dev work, but let's remove before merging

File: arctic_training/trainer/dpo_trainer.py [hunk 18]
```diff
@@ -0,0 +1,324 @@
...
+from arctic_training.model.hf_factory import HFModelFactory
+from arctic_training.model.liger_factory import LigerModelFactory
+from arctic_training.optimizer.adam_factory import FusedAdamOptimizerFactory
+from arctic_training.registry import get_registered_model_factory
+from arctic_training.scheduler.hf_factory import HFSchedulerFactory
+from arctic_training.tokenizer.hf_factory import HFTokenizerFactory
+from arctic_training.trainer.trainer import Trainer
+
+
+def to_device(batch: Dict, device: str) -> Dict:
```
Comment (sfc-gh-zhyao) [inline arctic_training/trainer/dpo_trainer.py:57]: @sfc-gh-mwyatt @sfc-gh-bzhai I think there is a common function for this? SFT needs it as well

File: arctic_training/data/dpo_factory.py [hunk 19]
```diff
@@ -0,0 +1,363 @@
...
+from torch.utils.data import DistributedSampler
+from transformers import BatchEncoding
+from transformers import PreTrainedTokenizerBase
+
+from arctic_training.config.data import DataConfig
+from arctic_training.data.factory import DataFactory
+from arctic_training.data.sft_factory import pad
+from arctic_training.data.utils import DatasetType
+
+IGNORE_INDEX = -100
```
Comment (sfc-gh-mwyatt) [inline arctic_training/data/dpo_factory.py:32]: There is a `IGNORE_INDEX` and `DPOTrainerConfig.ignore_label_index` we should use one or the other

File: arctic_training/trainer/dpo_trainer.py [hunk 20]
```diff
@@ -0,0 +1,324 @@
...
+        "zero_optimization": zero_opt_dict,
+        data_type: dtype_config,
+        "gradient_clipping": 1.0,
+        "prescale_gradients": False,
+        "wall_clock_breakdown": False,
+    }
+
+
+class DPOTrainerConfig(TrainerConfig):
+    ref_model: ModelConfig
```
Comment (sfc-gh-mwyatt) [inline arctic_training/trainer/dpo_trainer.py:121]: This needs a doc string to describe what is the field

File: projects/excot_dpo/data_generation/configs/bird_config.yaml [hunk 21]
```diff
@@ -0,0 +1,24 @@
+data:
+  bird:
+    train:
+      data_dir: "/data-fast/train"
```
Comment (sfc-gh-mwyatt) [inline projects/excot_dpo/data_generation/configs/bird_config.yaml:4]: Maybe update these `/data-fast/` paths to be more generic? `/tmp/` or `/data/` like in other examples

File: projects/excot_dpo/data_generation/configs/spider_config.yaml [hunk 22]
```diff
@@ -0,0 +1,18 @@
+spider:
+  train:
+    data_dir: "/data/bohan/spider"
```
Comment (sfc-gh-mwyatt) [inline projects/excot_dpo/data_generation/configs/spider_config.yaml:3]: remove `bohan` from paths

File: arctic_training/trainer/dpo_trainer.py (hunk 18 above)
Comment (sfc-gh-mwyatt) [reply arctic_training/trainer/dpo_trainer.py:57]: Add this in #116

File: arctic_training/data/dpo_factory.py (hunk 19 above)
Comment (sfc-gh-jrasley) [reply arctic_training/data/dpo_factory.py:33]: Will address in follow-up PR

**Review by sfc-gh-jrasley — APPROVED** [review]
